File: benchmark/haystack_8912/repro_script.py
import os
from unittest.mock import patch, MagicMock

# Import the necessary Haystack components
from haystack.components.generators.chat.openai import OpenAIChatGenerator
from haystack.dataclasses.chat_message import ChatMessage
from haystack.components.websearch import SerperDevWebSearch
from haystack.tools import ComponentTool

# --- Setup ---
# 1. Set dummy API keys to satisfy the __init__ methods of the real components.
#os.environ["OPENAI_API_KEY"] = "DUMMY_KEY"
#os.environ["SERPERDEV_API_KEY"] = "DUMMY_KEY"

print("--- Initializing OpenAIChatGenerator with tools_strict=True ---")

# 2. Instantiate the real components needed for the test.
web_search_component = SerperDevWebSearch()
search_tool = ComponentTool(component=web_search_component, name="web_search")

# 3. Initialize the generator with the strict tools flag enabled.
generator = OpenAIChatGenerator(model="gpt-4o", tools=[search_tool], tools_strict=True)

# 4. Define the target for our mock: the generator's internal method for preparing the API call.
#    This allows us to inspect the generated data without running the full API call cycle.
patch_target = "haystack.components.generators.chat.openai.OpenAIChatGenerator._prepare_api_call"


# --- Verification Logic ---
print("--- Calling generator.run() to trigger internal schema creation. ---")

try:
    # We patch the generator's internal method.
    with patch(patch_target, new_callable=MagicMock) as mock_prepare_api_call:
        # We need the mock to return a valid dictionary so the rest of the
        # `run` method doesn't crash, even though we will exit before it finishes.
        mock_prepare_api_call.return_value = {"messages": [], "model": "gpt-4o"}

        # This call will now trigger our mock.
        generator.run(messages=[ChatMessage.from_user("some query")])

        # --- Verification of the Bug ---
        # Get the keyword arguments that were passed to our mock.
        if not mock_prepare_api_call.called:
             raise AssertionError("The generator's _prepare_api_call method was not called.")
        
        # The arguments passed to our mock are what we need to inspect.
        # The full tool payload is generated and passed within these arguments.
        call_kwargs = mock_prepare_api_call.call_args.kwargs
        
        # The tool schemas are in the 'tools' keyword argument.
        tools_payload = call_kwargs.get("tools", [])
        
        if not tools_payload:
            raise AssertionError("The 'tools' payload was not generated and passed to _prepare_api_call.")

        # Find the function parameters for our specific tool
        tool_function_params = tools_payload[0].get("function", {}).get("parameters", {})

        print("\n--- Analyzing the generated tool schema ---")
        print(f"Generated Schema: {tool_function_params}")

        # The bug is that 'additionalProperties' is MISSING from the schema.
        if "additionalProperties" not in tool_function_params:
            print("\nSUCCESS: The bug is reproduced.")
            print("The 'additionalProperties' key was NOT found in the tool's function schema.")
            exit(1)
        else:
            print("\nFAILURE: The bug was NOT reproduced.")
            print("The 'additionalProperties' key was found in the schema.")
            exit(0)

except Exception as e:
    print(f"\nFAILURE: The script failed with an unexpected error: {type(e).__name__}: {e}")
    exit(0)

File: benchmark/ai_5628/repro_script.js
import { generateText, tool } from 'ai';
import { z } from 'zod';

// --- Mocking and Setup ---

// This is a mock of the Vercel AI SDK's LanguageModelV1 interface.
const mockVertexLanguageModel = {
  customAIProvider: undefined,

  // The doGenerate method is what `generateText` calls internally.
  // We will use it to capture and inspect the generated tool schema.
  doGenerate: async ({ tools }) => {
    console.log('--- Mock `doGenerate` called. Analyzing generated tools schema. ---');

    // Isolate the specific part of the schema that is buggy.
    const testToolSchema = tools?.find(t => t.function.name === 'test');
    const somethingProperty = testToolSchema?.function.parameters.properties.something;

    console.log('\nGenerated schema for the "something" property:');
    console.log(JSON.stringify(somethingProperty, null, 2));

    // --- Verification of the Bug ---
    // The bug is that for a `z.record(z.string())`, the generated schema
    // is either undefined or missing the required `"additionalProperties"` field.
    // This corrected check handles both cases.

    if (!somethingProperty || !('additionalProperties' in somethingProperty)) {
        console.log("\nSUCCESS: The bug is reproduced.");
        if (somethingProperty === undefined) {
            console.log("The generated schema for the 'something' property was 'undefined'.");
        } else {
            console.log("The 'additionalProperties' key was NOT found in the generated schema.");
        }
        // Exit with a non-zero code to signal to the test runner that the bug was found.
        process.exit(1);
    } else {
        console.log("\nFAILURE: The bug was NOT reproduced.");
        console.log("The 'additionalProperties' key was found in the schema, indicating the code is fixed.");
        process.exit(0);
    }

    // This part is just to satisfy the return type of doGenerate.
    return {
      toolCalls: [],
      finishReason: 'stop',
    };
  },
};


// --- Main Test Logic ---

async function runTest() {
  console.log('--- Calling generateText with a buggy z.record() schema. ---');
  try {
    await generateText({
      model: mockVertexLanguageModel,
      prompt: 'Hi',
      tools: {
        test: tool({
          description: 'Test',
          parameters: z.object({
            // This is the specific schema that causes the bug.
            something: z.record(z.string()),
          }),
        }),
      },
    });
  } catch (error) {
    console.error("\nFAILURE: The script failed with an unexpected error:", error);
    process.exit(0);
  }
}

// Run the asynchronous test function.
runTest();
